Refactor Add to App Android (#9297)
flutter/website: src/add-to-app/android/project-setup.md

  the Flutter module still runs correctly.
{{site.alert.end}}

## Integrate your Flutter module

{% comment %} Nav tabs {% endcomment -%}
<ul class="nav nav-tabs" id="add-to-app-android" role="tablist">
    <li class="nav-item">
        <a class="nav-link active" id="with-android-studio-tab" href="#with-android-studio" role="tab" aria-controls="with-android-studio" aria-selected="true">With Android Studio</a>
    </li>
    <li class="nav-item">
        <a class="nav-link" id="without-android-studio-tab" href="#without-android-studio" role="tab" aria-controls="without-android-studio" aria-selected="false">Without Android Studio</a>
    </li>
</ul>

{% comment %} Tab panes {% endcomment -%}
<div class="tab-content">

<div class="tab-pane active" id="with-android-studio" role="tabpanel" aria-labelledby="with-android-studio-tab" markdown="1">

### Integrate with Android Studio
{:.no_toc}

The Android Studio IDE can help integrate your Flutter module.
Using Android Studio, you can edit both your Android and Flutter code
in the same IDE.

You can also use your IntelliJ Flutter plugin functionality like
Dart code completion, hot reload, and widget inspector.

> Android Studio supports add-to-app flows on Android Studio 2022.1 or later

Reviewer: Since in our setup instructions (https://docs.flutter.dev/get-started/editor?tab=androidstudio#:~:text=version%202022.2%20(Flamingo)%20or%20later) we say we require 2022.2 or later, I think we could just drop this. Or say "Android Studio 2022.2 or later is required" to match that earlier requirement.

Author: I can do that. We will be swapping in a variable after the PoC is done to say latest stable.

> Android Studio supports add-to-app flows on Android Studio 2022.1 or later

Author: **nitpick:** Change version.
```suggestion
Android Studio supports add-to-app flows on Android Studio 2022.2 or later
```

> with version 42 or later of the [Flutter plugin][] for IntelliJ.

Reviewer: We can definitely drop this mention, since version 42 is from 2019 and only compatible with 2019.x versions of IntelliJ/AS.

> with version 42 or later of the [Flutter plugin][] for IntelliJ.

Author: **chore:** Update number.
```suggestion
with version 75.1.1 or later of the [Flutter plugin][] for IntelliJ.
```

> The Android Studio integration requires a source code Gradle subproject.

Reviewer: I think this changes the meaning of the old text. I think the old text is not trying to say that a "(pre-existing) source code Gradle subproject is required" but rather that's how the module will be integrated.

Reviewer: Nice fix!

> The Android Studio integration requires a source code Gradle subproject.

Author: **suggestion:** Change the content.
```suggestion
To build your app, the Android Studio plugin configures your
Android project to add your Flutter module as a dependency.
```

1. Open your Android project in Android Studio.

1. Go to **File** > **New** > **New Project...**.
    The **New Project** dialog displays.

1. Click **Flutter**.

1. If asked to provide your **Flutter SDK path**, do so and click **Next**.

1. Complete the configuration of your Flutter module.

    * If you have an existing project:

        {: type="a"}

>         1. To choose an existing project, click **...**.
>         to the right of the **Project location** box.

Reviewer: Looks like this extra period isn't intended:

```suggestion
        1. To choose an existing project, click **...**
           to the right of the **Project location** box.
```

  the Flutter module still runs correctly.
{{site.alert.end}}

## Integrate your Flutter module

> {% comment %} Nav tabs {% endcomment -%}

Reviewer 2: The section feels like a large formatting departure from the existing md file. Can this be written in a more plain markdown style?

Author: We use tabs elsewhere when a distinct choice is offered and it would reduce page length. This applies. The only difference is the tabs.

<ul class="nav nav-tabs" id="add-to-app-android" role="tablist">
    <li class="nav-item">
        <a class="nav-link active" id="with-android-studio-tab" href="#with-android-studio" role="tab" aria-controls="with-android-studio" aria-selected="true">With Android Studio</a>
    </li>
    <li class="nav-item">
        <a class="nav-link" id="without-android-studio-tab" href="#without-android-studio" role="tab" aria-controls="without-android-studio" aria-selected="false">Without Android Studio</a>
    </li>
</ul>

{% comment %} Tab panes {% endcomment -%}
<div class="tab-content">

<div class="tab-pane active" id="with-android-studio" role="tabpanel" aria-labelledby="with-android-studio-tab" markdown="1">

### Integrate with Android Studio
{:.no_toc}

The Android Studio IDE can help integrate your Flutter module.
Using Android Studio, you can edit both your Android and Flutter code
in the same IDE.

> You can also use your IntelliJ Flutter plugin functionality like

Reviewer: ```suggestion
You can also use IntelliJ Flutter plugin functionality like
```

Dart code completion, hot reload, and widget inspector.

Android Studio supports add-to-app flows on Android Studio 2022.2 or later

> with version 75.1.1 or later of the [Flutter plugin][] for IntelliJ.

Reviewer: ```suggestion
with the [Flutter plugin][] for IntelliJ.
```

{{site.alert.end}}

{{site.alert.note}}
  To avoid Dex merging issues, `flutter.androidPackage` should
  not be identical to your host app's package name.
{{site.alert.end}}

#### Java version requirement
{:.no_toc}

> The Flutter Android Gradle plugin uses features found in Java 11 or later.

Reviewer: While it is technically related to Gradle in this context, I believe it is more correct to say the "Flutter Android engine" here since it uses the features as well. Could we be less specific and just say something like:

```suggestion
Flutter requires your project to declare compatibility with Java 11 or later.
```

> Before attempting to connect your Flutter module project 

Reviewer: ```suggestion
Before attempting to connect your Flutter module project
```

>   To avoid Dex merging issues, `flutter.androidPackage` should not be identical to your host app's package name

Reviewer: ```suggestion
  To avoid Dex merging issues, `flutter.androidPackage` should
  not be identical to your host app's package name.
```

{{site.alert.end}}

#### Java 11 requirement
{:.no_toc}

> The Flutter Android Gradle plugin uses Java 11 features.

Reviewer: Do you have more context on this change?

Are the specifics important anyway? Could we just say "Flutter requires Java 11 compatibility."?

Reviewer: If this was about downloading Java, I think it wouldn't be necessary, but this is about a configuration option during Android app setup. Many Android apps that would be investigating add-to-app may still have these set to 1.8 (Java 8) since Android was a bit behind the times for a while.

Author: According to [our source code](https://github.com/flutter/flutter/blob/main/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart#L1101-L1116), the Gradle plugin requires Java 11 exactly.

Reviewer 2: https://github.com/flutter/flutter/blob/72b69f944915e8aae2af6a5475d4d6d4f4a4e4af/packages/flutter_tools/lib/src/android/gradle_errors.dart#L631

I think this is the code being tested and in that we are saying java 11 or higher.

  To avoid Dex merging issues, `flutter.androidPackage` should not be identical to your host app's package name
{{site.alert.end}}

#### Java 11 requirement
{:.no_toc}

> The Flutter Android engine uses Java 11 features.

Author: According to [our source code](https://github.com/flutter/flutter/blob/main/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart#L1101-L1116), yes, it's Java 11.

> The Flutter Android engine uses Java 11 features.

Reviewer 3: My point is that we don't really need to mention a version of Java that is years old.

> The Flutter Android engine uses Java 11 features.

Author: My point is that we need to explicitly require Java 11. We can't be ambiguous. Java 12 won't work. Java 20 won't work. It has to be Java 11.

one-click build process, but requires the Flutter SDK.
This is the mechanism used by the Android Studio IDE plugin.

> ### Option A - Depend on the Android Archive (AAR)

Reviewer 3: Shouldn't "Option A and Option B" use another set of 2 tabs? It seems odd to diverge from tabs after using them above.

Also, I don't see the text here, but there are a couple of Notes in the staged site that talk about .android and .androidPackage files. I assume these notes only apply to Mac. What about Windows?

Author: 1. Changed to tabs.
2. The `.android` directory is a local version of the full Android OS for the emulator. It exists in Windows, too.

Reviewer 3: Interesting. Dot files are generally a Unix-only convention.
